modules/codegen/src/main/scala/com/twilio/guardrail/generators/CirceProtocolGenerator.scala

```diff
@@ -356,4 +383,62 @@ object CirceProtocolGenerator {
         """.stats)
     }
   }
+
+  object PolyProtocolTermInterp extends (PolyProtocolTerm ~> Target) {
+    override def apply[A](fa: PolyProtocolTerm[A]): Target[A] = fa match {
+      case RenderADTCompanion(clsName, discriminator, encoder, decoder) =>
+        val code = q"""object ${Term.Name(clsName)} {
+             val discriminator:String = ${Lit.String(discriminator)}
+            ..${List(encoder, decoder)}
+          }
+          """
+        Target.pure(code)
+
+      case DecodeADT(clsName, children) =>
+        val childrenCases = children.map(
+          child => p"case ${Lit.String(child)} => c.as[${Type.Name(child)}]"
+        )
+        val code =
+          q"""implicit val decoder: Decoder[${Type.Name(clsName)}] = Decoder.instance(c =>
+                 c.downField(discriminator).as[String].flatMap {
+                   ..case $childrenCases
+                 }
+            )"""
+        Target.pure(code)
+
+      case EncodeADT(clsName, children) =>
+        val childrenCases = children.map(
+          child => p"case e:${Type.Name(child)} => e.asJsonObject.add(discriminator, ${Lit.String(child)}.asJson).asJson"
+        )
+        val code =
+          q"""implicit val encoder: Encoder[${Type.Name(clsName)}] = Encoder.instance {
+              ..case $childrenCases
+          }"""
+        Target.pure(code)
+
+      case RenderSealedTrait(className, terms, discriminator, parents) =>
+        val parentNameOpt = parents.headOption.map(_.clsName)
+        val testTerms = terms
+          .filter(_.name.value != discriminator)
+          .map { t =>
+            val tpe: Type = t.decltpe
+              .flatMap({
+                case tpe: Type => Some(tpe)
+                case x =>
+                  println(s"Unsure how to map ${x.structure}, please report this bug!")
+                  None
+              })
+              .get
```

Comment:
```scala
      case RenderSealedTrait(className, terms, discriminator, parents) =>
        for {
          testTerms <- (
            terms
              .filter(_.name.value != discriminator)
              .traverse { t =>
                for {
                  tpe <- Target.fromOption(
                    t.decltpe
                      .flatMap({
                        case tpe: Type => Some(tpe)
                        case x => None
                      })
                    , s"Unsure how to map ${x.structure}, please report this bug!")
                } yield q"""def ${Term.Name(t.name.value)}: ${tpe}"""
              }
          )
        } yield {
          parents
            .headOption
            .fold(q"""trait ${Type.Name(className)} {..${testTerms}}""")(
              parent => q"""trait ${Type.Name(className)} extends ${template"${init"${Type.Name(parent.clsName)}(...$Nil)"}{..${testTerms}}"} """
            )
        }
```
please

Reply:
Thank you so much. I've only changed one line.
`
- s"Unsure how to map ${x.structure}, please report this bug!")
+ t.decltpe.fold("Nothing to map")(x => s"Unsure how to map ${x.structure}, please report this bug!"))
`